app/modules/ws2812_effects.c

```diff
@@ -0,0 +1,1074 @@
+{
+
+  if (state->effect_type == WS2812_EFFECT_BLINK)
+  {
+    ws2812_effects_mode_blink();
+  }
+  else if (state->effect_type == WS2812_EFFECT_RAINBOW)
+  {
+    ws2812_effects_mode_rainbow();
+  }
+  else if (state->effect_type == WS2812_EFFECT_RAINBOW_CYCLE)
+  {
+    // the rainbow cycle effect can be achieved by shifting the buffer
+    ws2812_buffer_shift(state->buffer, 1, SHIFT_CIRCULAR, 1, -1);
+  }
+  else if (state->effect_type == WS2812_EFFECT_FLICKER)
+  {
+    int flicker_value = state->effect_int_param1 != EFFECT_PARAM_INVALID ? state->effect_int_param1 : 100;
+    if (flicker_value == 0) {
+      flicker_value = 50;
+    }
+    ws2812_effects_mode_flicker_int(flicker_value);
+    state->counter_mode_step = (state->counter_mode_step + 1) % 256;
+  }
+  else if (state->effect_type == WS2812_EFFECT_FIRE_FLICKER)
+  {
+    ws2812_effects_mode_flicker_int(110);
+    state->counter_mode_step = (state->counter_mode_step + 1) % 256;
+  }
+  else if (state->effect_type == WS2812_EFFECT_FIRE_FLICKER_SOFT)
+  {
+    ws2812_effects_mode_flicker_int(70);
+    state->counter_mode_step = (state->counter_mode_step + 1) % 256;
+  }
+  else if (state->effect_type == WS2812_EFFECT_FIRE_FLICKER_INTENSE)
+  {
+    ws2812_effects_mode_flicker_int(170);
+    state->counter_mode_step = (state->counter_mode_step + 1) % 256;
+  }
+  else if (state->effect_type == WS2812_EFFECT_RANDOM_COLOR)
+  {
+    ws2812_effects_mode_random_color();
+  }
+  else if (state->effect_type == WS2812_EFFECT_HALLOWEEN)
+  {
+    ws2812_buffer_shift(state->buffer, 1, SHIFT_CIRCULAR, 1, -1);
+  }
+  else if (state->effect_type == WS2812_EFFECT_CIRCUS_COMBUSTUS)
+  {
+    ws2812_buffer_shift(state->buffer, 1, SHIFT_CIRCULAR, 1, -1);
+  }
+  else if (state->effect_type == WS2812_EFFECT_LARSON_SCANNER)
+  {
+    ws2812_effects_mode_larson_scanner();
+  }
+  else if (state->effect_type == WS2812_EFFECT_CYCLE)
+  {
+    ws2812_buffer_shift(state->buffer, state->effect_int_param1, SHIFT_CIRCULAR, 1, -1);
+  }
+  else if (state->effect_type == WS2812_EFFECT_COLOR_WIPE)
+  {
+    ws2812_effects_mode_color_wipe();
+  }
+  else if (state->effect_type == WS2812_EFFECT_RANDOM_DOT)
+  {
+    uint8_t dots = state->effect_int_param1 != EFFECT_PARAM_INVALID ? state->effect_int_param1 : 1;
+    ws2812_effects_mode_random_dot(dots);
+  }
+
+  // set the new delay for this effect
+  state->mode_delay = ws2812_effects_mode_delay();
+  // call count
+  state->counter_mode_call = (state->counter_mode_call + 1) % UINT32_MAX;
+  // write the buffer
+  ws2812_write(state->buffer);
+  // set the timer
+  if (state->running == 1 && state->mode_delay >= 10)
+  {
+    os_timer_disarm(&(state->os_t));
+    os_timer_arm(&(state->os_t), state->mode_delay, FALSE);
+  }
+}
+
+
+/**
+* Set the active effect mode
+*/
+static int ws2812_effects_set_mode(lua_State* L) {
+
+  luaL_argcheck(L, state != NULL, 1, LIBRARY_NOT_INITIALIZED_ERROR_MSG);
+  const int mode = luaL_checkinteger(L, 1);
```

Comment:
Quoting [`luaL_checkoption()`](http://pgl.yoyo.org/luai/i/luaL_checkoption):
> The usual convention in Lua libraries is to use strings instead of numbers to select options.

I know that our firmware is not 100% consistent here, but please think about turning the mode argument into string for the above reason. See `file`, `gpio` and other modules for how to use it.

Reply:
I am fine with adjusting this, I just "cloned" the way it is done in the ws2812 module itself (for fading and single/dual mode). This was the closest thing to look at